File: 1995-finding-pairs-with-a-certain-sum/1995-finding-pairs-with-a-certain-sum.go
type FindSumPairs struct {
    mp      map[int]int
    nums1   []int
    nums2   []int
}


func Constructor(nums1 []int, nums2 []int) FindSumPairs {
    mp := make(map[int]int)
    for _, num := range nums2 {
        _, valid := mp[num]
        if valid {
            mp[num]++
        } else {
            mp[num] = 1
        }
    }
    return FindSumPairs{mp, nums1, nums2}
}


func (this *FindSumPairs) Add(index int, val int)  {
    prev := this.nums2[index]
    this.mp[prev]--
    this.mp[prev + val]++
    this.nums2[index] = prev + val
}


func (this *FindSumPairs) Count(tot int) int {
    ans := 0
    for _, num := range this.nums1 {
        f, present := this.mp[tot - num]
        if present {
            ans += f
        }
    }
    return ans
}


/**
 * Your FindSumPairs object will be instantiated and called as such:
 * obj := Constructor(nums1, nums2);
 * obj.Add(index,val);
 * param_2 := obj.Count(tot);
 */

File: 1995-finding-pairs-with-a-certain-sum/1995-finding-pairs-with-a-certain-sum.cpp
class FindSumPairs {
private:
    unordered_map<int, int> mp;
    vector<int> nums1, nums2;

public:
    FindSumPairs(vector<int>& nums1, vector<int>& nums2) {
        this->nums1 = nums1;
        this->nums2 = nums2;
        for(int &num : nums2) this->mp[num]++;
    }
    
    void add(int index, int val) {
        int prev = nums2[index];
        mp[prev]--;
        mp[prev + val]++;
        nums2[index] = prev + val;
    }
    
    int count(int tot) {
        int ans = 0;
        for(int &num : nums1) if(mp.count(tot - num)) ans += mp[tot - num];
        return ans;
    }
};

/**
 * Your FindSumPairs object will be instantiated and called as such:
 * FindSumPairs* obj = new FindSumPairs(nums1, nums2);
 * obj->add(index,val);
 * int param_2 = obj->count(tot);
 */